Decode with mem limit

serban300 opened the pull request:
Related to https://github.com/paritytech/parity-scale-codec/issues/609

gui1117 commented:
The current implementation is enough for the usage or should we also implement it automatically in derive macro?

gui1117 commented on src/codec.rs:
```diff
@@ -466,17 +486,29 @@ where
 	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
-		input.scale_internal_decode_bytes()
+		let len = <Compact<u32>>::decode(input)?.0 as usize;
+		if input.__private_bytes_cursor().is_some() {
+			input.on_before_alloc_mem(len)?;
+		}
+
+		if let Some(bytes_cursor) = input.__private_bytes_cursor() {
+			bytes_cursor.decode_bytes_with_len(len)
+		} else {
```
```suggestion
		if let Some(bytes_cursor) = input.__private_bytes_cursor() {
			input.on_before_alloc_mem(len)?;
			bytes_cursor.decode_bytes_with_len(len)
		} else {
```

is it more readable?

gui1117 replied:
ah indeed, my bad

gui1117 commented on src/codec.rs:
```diff
@@ -1207,13 +1274,20 @@ macro_rules! impl_codec_through_iterator {
 					input.descend_ref()?;
-					let result = Result::from_iter((0..len).map(|_| Decode::decode(input)));
+					let result = Result::from_iter((0..len).map(|_| {
+						input.on_before_alloc_mem(0 $( + mem::size_of::<$generics>() )*)?;
```
This is an underestimate of the allocation, is it fine?

All the collections below will use more space than what is reported.

maybe it should be documented in `Input` that it is a rough estimation then.

serban300 replied:
Yes, this is as desired. It is acceptable to be a bit off. The main thing is to track the memory used by the decoded data inside the type.

You're right, it's good to document this. Added a comment in `Input`.

gui1117 replied:
great :+1:

gui1117 commented on src/codec.rs:
```diff
@@ -1260,6 +1334,8 @@ impl<T: Decode> Decode for VecDeque<T> {
+
 impl EncodeLike for () {}
 
 impl Encode for () {
```
It seems `DecodeWithMemTracking` is missed for tuples and for `()`.

gui1117 replied:
ah I missed it, no it is good as it is.

gui1117 reviewed (commented):
overall looks good,

gui1117 commented:
> Yes, it would be nice to also implement it automatically. Normally it should be an auto trait, but this feature is still unstable so I couldn't use that.
> 
> Implementing it automatically in the derive macros would be the next best thing, but I didn't manage to do it. For example for:
> 
> ```
> struct Test 
> {
>     a: u32
> }
> ```
> 
> I tried something that would expand to:
> 
> ```
> impl DecodeWithMemTracking for Test where u32: DecodeWithMemTracking {
> }
> ```
> 
> Which generates compile errors.
> 
> So probably we'll need to implement it manually for each custom struct.

I see we can't derive it automatically with `Decode` derive-macro, but we could provide a `DecodeWithMem` derive-macro.

This macro will do the bound same as Encode/Decode, i.e. bound only the generic types. and it will fail if some concrete types don't implement `DecodeWithMemTracking`.

It would still be more handy than manually write the code I guess.

gui1117 commented on src/mem_tracking.rs:
```diff
@@ -0,0 +1,86 @@
+
+/// Marker trait used for identifying types that call the mem tracking hooks exposed by `Input`
+/// while decoding.
+pub trait DecodeWithMemTracking: Decode {}
```
I guess it would be great to implement it also on:
* `Compact<T>` where `Compact<T::As>: DecodeWithMemTracking`
* all the explicit compact: `Compact<()>`, `Compact<u8>`...`Compact<u128>`
* stuff in bit_vec module
* stuff in generic_array module

serban300 replied:
Done for all except `generic_array` with no additional changes needed.

With `generic_array` I'm not familiar and I don't know if anyone uses it. In polkadot-sdk it doesn't seem to be used. We could implement it later if someone requested it.

gui1117 replied:
Yes we can implement generic_array on demand.

gui1117 commented on tests/mem_tracking.rs:
```diff
@@ -0,0 +1,132 @@
+		1015
+	)
+	.is_ok())
+}
```
maybe one more test where the decoding fails because mem limit is reached.

serban300 replied:
Good point ! Done.

gui1117 reviewed (commented):
I suggested some more types to implement the trait on.

But part from this it looks good to me.

serban300 commented:
> I see we can't derive it automatically with `Decode` derive-macro, but we could provide a `DecodeWithMem` derive-macro.
> 
> This macro will do the bound same as Encode/Decode, i.e. bound only the generic types. and it will fail if some concrete types don't implement `DecodeWithMemTracking`.
> 
> It would still be more handy than manually write the code I guess.

I'm not sure if it's possible. Unless we add a mock method to the `DecodeWithMemTracking` trait and try to call it in the derived implementation. Which would be a bit of a hack. But even so, implementing it manually isn't so hard. It's just an empty marker trait. I'm not sure if deriving it would make things much easier.

ggwpez commented on src/codec.rs:
```diff
@@ -408,12 +420,32 @@ mod feature_wrapper_bytes {
+/// `Input` implementation optimized for decoding `bytes::Bytes`.
 #[cfg(feature = "bytes")]
-struct BytesCursor {
+pub struct BytesCursor {
```
Why are you making it pub? Then we have to include it our public SemVer API.

serban300 replied:
Reverted the change. Not applicable anymore.

gui1117 commented:
> > I see we can't derive it automatically with `Decode` derive-macro, but we could provide a `DecodeWithMem` derive-macro.
> > This macro will do the bound same as Encode/Decode, i.e. bound only the generic types. and it will fail if some concrete types don't implement `DecodeWithMemTracking`.
> > It would still be more handy than manually write the code I guess.
> 
> I'm not sure if it's possible. Unless we add a mock method to the `DecodeWithMemTracking` trait and try to call it in the derived implementation. Which would be a bit of a hack. But even so, implementing it manually isn't so hard. It's just an empty marker trait. I'm not sure if deriving it would make things much easier.

I see it now, indeed.

Then ok

ggwpez commented on src/codec.rs:
```diff
@@ -1207,13 +1280,20 @@ macro_rules! impl_codec_through_iterator {
 					input.descend_ref()?;
-					let result = Result::from_iter((0..len).map(|_| Decode::decode(input)));
+					let result = Result::from_iter((0..len).map(|_| {
+						input.on_before_alloc_mem(0 $( + mem::size_of::<$generics>() )*)?;
```
Since this is low level runtime code, we need to be extremely paranoid about panics. Please use saturating math.

bkchr replied:
`mem::size_of` is a constant function, so it should already fail at compile time.

serban300 replied:
Changed this. Not applicable anymore.

ggwpez commented on src/mem_tracking.rs:
```diff
@@ -0,0 +1,86 @@
+
+/// Marker trait used for identifying types that call the mem tracking hooks exposed by `Input`
+/// while decoding.
+pub trait DecodeWithMemTracking: Decode {}
```
Can you mention what hook it calls please?

ggwpez commented on tests/mem_tracking.rs:
```diff
@@ -0,0 +1,158 @@
+
+impl DecodeWithMemTracking for ComplexStruct {}
+
+fn decode_object<T>(obj: T, mem_limit: usize, expected_used_mem: usize) -> Result<T, Error>
```
This function could also call the decode again with one byte less memory limit to check that it fails.  
Then you get all the error cases also covered.

bkchr reviewed (commented):
Did you try this will work on Substrate? I mean you will need to ensure that `Call` implements `DecodeWithMemTracking` and all the types we are using there.

Also you should expose a function `decode_with_mem_limit` instead of exposing the `MemLimitInput`

serban300 commented:
> Did you try this will work on Substrate? I mean you will need to ensure that `Call` implements `DecodeWithMemTracking` and all the types we are using there.

Oh, you're right. I can patch polkadot-sdk locally and try this. Will do it and get back.

> Also you should expose a function `decode_with_mem_limit` instead of exposing the `MemLimitInput`

Yes, will do this too.

bkchr commented on src/codec.rs:
```diff
@@ -85,16 +86,27 @@ pub trait Input {
 	#[cfg(feature = "bytes")]
 	#[doc(hidden)]
-	fn scale_internal_decode_bytes(&mut self) -> Result<bytes::Bytes, Error>
+	fn __private_bytes_cursor(&mut self) -> Option<&mut BytesCursor>
```
What you are saying there makes no real sense. Decoding `Bytes` from bytes currently doesn't work with any other input. This is by design. So, the changes you have done here are not really required.



> The same applies for combining `BytesCursor` with `DepthLimit` or anything else.

You can simply not do this right now. `BytesCursor` is right now not exposed to the outside and thus, you can also not use it as `input` for `DepthLimit`.

In the end what you would need to do is that you forward the `scale_internal_decode_bytes` call in `DepthLimit` to the inner input for example. Still, this would require that `BytesCursor` is public, which we should not do right now.

serban300 replied:
Hmmm yes, makes sense. Because `BytesCursor` is private. I didn't think of that. Ok, reverted this change then.

bkchr commented on src/codec.rs:
```diff
@@ -1207,13 +1258,20 @@ macro_rules! impl_codec_through_iterator {
 				<Compact<u32>>::decode(input).and_then(move |Compact(len)| {
 					input.descend_ref()?;
-					let result = Result::from_iter((0..len).map(|_| Decode::decode(input)));
+					let result = Result::from_iter((0..len).map(|_| {
```
Most of the time containers are implemented to always increase the capacity by two when hitting the limit. This means here that we may only decode 190 elements while the capacity is already at 256. What I want to say is that you will not track the size correctly.

serban300 replied:
True. Initially I wanted to avoid the complexity of keeping track of this. I saw that other libraries did the same. For example the `heapsize` lib from `servo`: https://github.com/servo/heapsize/blob/master/src/lib.rs#L295 . But anyway, done. PTAL !

serban300 commented on src/codec.rs:
```diff
@@ -408,12 +420,32 @@ mod feature_wrapper_bytes {
+	}
+
+	fn decode_bytes_with_len(&mut self, length: usize) -> Result<bytes::Bytes, Error> {
+		bytes::Buf::advance(&mut self.bytes, self.position);
```
Reverted the change. Not applicable anymore.

serban300 commented:
> I still think requiring to implement it manually can be painful. But if the types requiring this are not many then it good.

Yes, sorry, you're right. We need to be able to derive it since for types like `RuntimeCall` we can't implement it manually. Done.

serban300 commented:
> > Did you try this will work on Substrate? I mean you will need to ensure that `Call` implements `DecodeWithMemTracking` and all the types we are using there.
> 
> Oh, you're right. I can patch `polkadot-sdk` locally and try this. Will do it and get back.

Done. This compiles for `rococo-runtime`: https://github.com/serban300/polkadot-sdk/tree/mem-limit-decode . It's very verbose but seems to be implemented for all the types that we need.

> > Also you should expose a function `decode_with_mem_limit` instead of exposing the `MemLimitInput`
> 
> Yes, will do this too.

Done

@bkchr Addressed all the comments. PTAL !

bkchr commented on src/codec.rs:
```diff
@@ -1223,17 +1261,99 @@ macro_rules! impl_codec_through_iterator {
+const BTREE_MIN_LEN_AFTER_SPLIT: usize = BTREE_B - 1;
+
+/// Estimate the mem size of a btree.
+fn mem_size_of_btree<T>(len: u32) -> usize {
```
Would be nice to have some sort of test for this, but also not that easy to implement. We would require some custom allocator to get the statistics.

bkchr replied:
But not a requirement for this pr

serban300 replied:
Good point ! Added a simple test and also adjusted `mem_size_of_btree()` a bit.

bkchr commented on derive/src/decode.rs:
```diff
@@ -300,3 +285,44 @@ fn create_instance(
+		} else {
+			field.ty.to_token_stream()
+		};
+		Some(quote! {<#field_type as #crate_path::DecodeWithMemTracking>::__is_implemented();})
```
This `__is_implemented` function should be removed and the checks should be put into the unnamed `const` and using something like this:
https://docs.rs/static_assertions/1.1.0/src/static_assertions/assert_impl.rs.html#57-83

(We should copy this and not import the crate) Generally we could probably also have the following code:
```
fn require_decode_with_mem::<T: DecodeWithMemTracking>() {}

require_decode_with_mem::<field_ty>()
```

This should probably also work.

serban300 replied:
Nice ! Done.

bkchr reviewed (commented):
Some last nitpicks. Otherwise it already looks good :)

serban300 commented on derive/src/utils.rs:
```diff
@@ -449,3 +467,16 @@ pub fn is_transparent(attrs: &[syn::Attribute]) -> bool {
+		))
+	}
+
+	Ok(data_variants().collect())
```
Yes, looks better. Done. Thanks !

serban300 commented on src/codec.rs:
```diff
@@ -1223,17 +1261,99 @@ macro_rules! impl_codec_through_iterator {
+			let result = Result::from_iter((0..len).map(|_| {
+				// We account for the size of the `prev` and `next` pointers of each list node,
+				// plus the decoded element.
+				input.on_before_alloc_mem(mem::size_of::<(usize, usize, T)>())?;
```
Good point. Done.

bkchr commented on src/btree_utils.rs:
```diff
@@ -0,0 +1,109 @@
+	num_nodes.saturating_mul(internal_node_size)
+}
+
+#[cfg(all(test, feature = "unstable-tests"))]
```
```suggestion
#[cfg(test)]
#[rustversion::nightly]
```

https://crates.io/crates/rustversion

This way we don't need to expose an feature for this.

serban300 replied:
Nice, but also had to add a build script to make it work, as stated here: https://github.com/dtolnay/rustversion/issues/8

PTAL !

ggwpez commented on derive/src/decode.rs:
```diff
@@ -300,3 +285,48 @@ fn create_instance(
+
+			let mut fields: Box<dyn Iterator<Item = &Field>> = Box::new(iter::empty());
+			for variant in variants {
+				fields = Box::new(fields.chain(variant.fields.iter()));
```
Why do you use an iter here instead of a vector?

serban300 replied:
It should be more efficient. If we want to build a vector we have to iterate through the fields and `collect()`. This way we just get the iterator and use it below.

ggwpez replied:
Huh, derive macros dont have to be optimized, but okay.

ggwpez commented on src/compact.rs:
```diff
@@ -249,7 +256,12 @@ impl<T> MaybeMaxEncodedLen for T {}
-	type Type: for<'a> EncodeAsRef<'a, Self> + Decode + From<Self> + Into<Self> + MaybeMaxEncodedLen;
+	type Type: for<'a> EncodeAsRef<'a, Self>
+		+ Decode
+		+ DecodeWithMemTracking
```
So only mem tracked types can be compact now? I guess that is fine, since `Compact` should be sealed anyway.

serban300 replied:
Yes, this was also my reasoning.

ggwpez commented on tests/mem_tracking.rs:
```diff
@@ -0,0 +1,153 @@
+	assert_eq!(input.used_mem(), expected_used_mem);
+	if expected_used_mem > 0 {
+		let raw_input = &mut &encoded_bytes[..];
+		let mut input = MemTrackingInput::new(raw_input, expected_used_mem);
```
Why does it fail with this? I thought that it should fail only with `expected_used_mem - 1`.

serban300 commented:
@bkchr just wanted to check if you would like to take another look on the PR or if I could merge it as it is since it already has 2 approving reviews

bkchr commented on src/compact.rs:
```diff
@@ -273,7 +285,12 @@ where
-	Compact<T>: for<'a> EncodeAsRef<'a, T> + Decode + From<Self> + Into<Self> + MaybeMaxEncodedLen,
+	Compact<T>: for<'a> EncodeAsRef<'a, T>
+		+ Decode
+		+ DecodeWithMemTracking
```
This and the change above are semver breaking and require a major bump.

serban300 replied:
Hmmm yes, you're right, good catch ! But, I wouldn't be very keen on bumping the major version, since then we would also have to implement https://github.com/paritytech/parity-scale-codec/issues?q=is%3Aopen+is%3Aissue+milestone%3A%22SCALE+codec+4.0%22 in order to do the release, and it would delay this feature. So WDYT about https://github.com/paritytech/parity-scale-codec/pull/616/commits/336c813bec1c27f4b413a536824ad46dc1be0865 as an alternative ? At least temporary until the other work for releasing 4.0 is done ?

serban300 replied:
@bkchr sorry for the ping. Can you take a look on the answer above please ?

bkchr commented on derive/src/decode.rs:
```diff
@@ -300,3 +285,48 @@ fn create_instance(
+		Some(quote! {#field_type})
+	});
+
+	quote! {
```
`quote_spanned` with the span of each type to get a better error. Maybe also a ui test would be nice

serban300 replied:
Done. I hope I understood correctly how `quote_spanned` works.

bkchr commented on derive/src/utils.rs:
```diff
@@ -85,17 +86,22 @@ pub fn get_encoded_as_type(field: &Field) -> Option<TokenStream> {
 }
 
 /// Look for a `#[codec(compact)]` outer attribute on the given `Field`.
```
Docs should be updated.

serban300 replied:
Done

bkchr commented on Cargo.toml:
```diff
@@ -1,24 +1,26 @@
 generic-array = { version = "0.14.7", optional = true }
 arbitrary = { version = "1.3.2", features = ["derive"], optional = true }
 impl-trait-for-tuples = "0.2.2"
+rustversion = "1.0.17"
```
We don't need it for the normal dependencies?

serban300 replied:
You're right. Missed it. Done.

bkchr reviewed (approved):
Nice work :+1:  Thank you for doing this!
